src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -360,24 +358,46 @@ private static int compareToCIImpl(byte[] value, int toffset, int tlen,
 
     // Case insensitive comparison of two code points
     private static int compareCodePointCI(int cp1, int cp2) {
+        if (latin1EqualsIgnoreCase(cp1, cp2)) {
```

Comment:
I don't think you can rule out all cross-latin1 case folds just with this check. There's a number of cases where codepoints fold across in either direction that we need to consider if one of the codepoints is above 255 and the other below or equal:
```java
jshell> for (int i = 256; i < 65536; i++) {
   ...>     char c = (char)i; if (Character.toLowerCase(c) < 256) {
   ...>         System.out.println("Char: " + c + " codepoint: " + i + "\tlowercase: " + Character.toLowerCase(c) + " lowercase codepoint: " + (int)Character.toLowerCase(c));
   ...> }
   ...> }
Char: İ codepoint: 304	lowercase: i lowercase codepoint: 105
Char: Ÿ codepoint: 376	lowercase: ÿ lowercase codepoint: 255
Char: ẞ codepoint: 7838	lowercase: ß lowercase codepoint: 223
Char: K codepoint: 8490	lowercase: k lowercase codepoint: 107
Char: Å codepoint: 8491	lowercase: å lowercase codepoint: 229

jshell> for (int i = 0; i < 256; i++) {
   ...>     char c = (char)i; if (Character.toUpperCase(c) > 255) {
   ...>         System.out.println("Char: " + c + " codepoint: " + i + "\tuppercase: " + Character.toUpperCase(c) + " uppercase codepoint: " + (int)Character.toUpperCase(c));
   ...>     }
   ...> }
Char: µ codepoint: 181	uppercase: Μ uppercase codepoint: 924
Char: ÿ codepoint: 255	uppercase: Ÿ uppercase codepoint: 376
```

Reverting the `ofUnicode` shortcut below might be sufficient to make this method correct, but probably hurts performance.

Reply:
I noticed the same so I reverted the CharacterData change. The performence is now neutral instead of a win.

Reply:
This might be an issue in the mixed-coder comparison as well. I'll investigate.